// @healerlab/vue3-simple-otp-input/dist/h-otp-input.js
import { defineComponent as k, watch as P, ref as g, onMounted as R, openBlock as f, createElementBlock as i, normalizeClass as C, Fragment as D, renderList as q, withDirectives as K, normalizeStyle as N, withModifiers as M, vModelDynamic as z } from "vue";
const E = ["element-num", "id", "onUpdate:modelValue", "type", "disabled", "readonly", "onKeyup", "onPaste", "onFocus"], U = /* @__PURE__ */ k({
  __name: "HOtpInput",
  props: {
    length: {
      type: Number,
      required: !0,
      default: 6
    },
    defaultColor: {
      type: String,
      default: "#eeeeee"
    },
    mainColor: {
      type: String,
      default: "#00dc82",
      required: !1
    },
    fontColor: {
      type: String,
      default: "black",
      required: !1
    },
    allowPaste: {
      type: Boolean,
      default: !0,
      required: !1
    },
    onlyNumber: {
      type: Boolean,
      default: !1
    },
    isRefresh: {
      type: Boolean,
      default: !1
    },
    type: {
      type: String,
      default: "password",
      validator: (r) => ["password", "text"].includes(r)
    },
    autoFocus: {
      type: Boolean,
      default: !0
    },
    disabled: {
      type: Boolean,
      default: !1
    },
    outlined: {
      type: Boolean,
      default: !1
    },
    readonly: {
      type: Boolean,
      default: !1
    },
    wrapClass: {
      type: String,
      default: ""
    },
    inputClass: {
      type: String,
      default: ""
    }
  },
  emits: ["onComplete", "onChange"],
  setup(r, { expose: h, emit: p }) {
    const o = r;
    P(
      () => o.isRefresh,
      () => {
        a.value = [], l[0].focus(), s.value = -1;
      }
    );
    const l = {}, a = g([]), s = g(-1);
    h({
      clear: () => {
        a.value = [], l[0].focus(), s.value = -1;
      }
    });
    const v = (e) => {
      if (!o.onlyNumber)
        return;
      const t = e.charCode || e.keyCode || 0;
      t === 8 || t === 46 || t === 86 || t === 91 || t >= 48 && t <= 57 || t >= 96 && t <= 105 || e.preventDefault();
    }, b = (e, t) => {
      l[e] = t;
    }, B = (e) => `hl-${e + 1}`, y = (e) => `3px solid ${a.value[e] ? o.mainColor : o.defaultColor}`, w = (e) => {
      a.value[e] && a.value[e] !== "" && e < o.length ? (s.value = e + 1, l[e + 1] && l[e + 1].focus()) : e > 0 && (!a.value[e] || a.value[e] === "") && l[e - 1] && (l[e - 1].focus(), s.value = e - 1);
    }, F = (e) => {
      l[e].select();
    }, I = (e) => {
      var t;
      if (o.allowPaste) {
        const d = ((t = e == null ? void 0 : e.clipboardData) == null ? void 0 : t.getData("text/plain")) || "", n = d.split("");
        let u = !0;
        if (o.onlyNumber && (u = new RegExp(`^\\d{${o.length}}$`).test(d)), u) {
          for (let c = 0; c < o.length; c++)
            S(c, n[c]);
          l[o.length - 1].focus(), m();
        }
      }
    }, S = (e, t) => {
      a.value[e] = t;
    }, m = () => {
      const e = a.value.join("");
      e.length === o.length ? (p("onComplete", e), s.value = -1) : p("onChange", e);
    };
    R(() => {
      o.autoFocus && l && l[0].focus();
    });
    const $ = (e) => o.outlined ? {
      outline: V(s.value === e),
      borderBottom: "none",
      borderRadius: "2px"
    } : {
      borderBottom: y(e)
    }, V = (e) => {
      const t = e ? o.mainColor : o.defaultColor;
      return o.outlined ? `2px solid ${t}` : "none";
    };
    return (e, t) => (f(), i("div", {
      class: C(["hl-input-container m-auto text-center", r.wrapClass])
    }, [
      (f(!0), i(D, null, q(r.length, (d, n) => K((f(), i("input", {
        "element-num": d,
        id: B(n),
        ref_for: !0,
        ref: (u) => b(n, u),
        key: n,
        "onUpdate:modelValue": (u) => a.value[n] = u,
        type: r.type,
        maxlength: "1",
        style: N({
          borderBottom: y(n),
          color: r.fontColor,
          ...$(n)
        }),
        class: C(r.inputClass),
        disabled: r.disabled,
        readonly: r.readonly,
        contenteditable: "true",
        onKeydown: t[0] || (t[0] = (u) => v(u)),
        onKeyup: (u) => w(n),
        onPaste: M(I, ["prevent"]),
        onInput: t[1] || (t[1] = (u) => m()),
        onFocus: (u) => F(n)
      }, null, 46, E)), [
        [z, a.value[n]]
      ])), 128))
    ], 2));
  }
});
export {
  U as HOtpInput
};
